### ISSUE TITLE
PHP Halo API - get___ requests that return null data run slowly

### ISSUE THREAD

**Issue by ActionCactus** [ff36/halo-core#83]
When running getSectors(), getNodes(), getProperties(), or getIncidents() against a Halo server the execution time is drastically slower (up to 60 times slower in my testing) for data sets that are NULL.

**Comment by tarkal** [comment]
Could you elaborate on a couple of points.

1. I know you are running the PHP SDK. Have you tested this against the raw HTTP requests or only through the SDK? Just trying to determine if the problem is originating from the core or if the PHP SDK is basically holding the request until its default timeout.

2. When you say `NULL` data sets, do you mean you are targeting a genuine endpoint that has an empty collection response or you are trying to target an unknown endpoint?

**Comment by ActionCactus** [comment]
1.  This has been replicated in Postman as well

2.  The following request may offer some insight:

http://URL/api/sectors/22:5/nodes/23:6592/properties?page=3&perPage=100

This takes several seconds to respond, whereas if you point to page 1 it's almost instantaneous.

**Comment by tarkal** [comment]
OK so using the `page=3&perPage=100` returns an empty collection?

**Comment by ActionCactus** [comment]
Indeed, it does.  As long as there aren't more than 200 properties for that node

**Comment by tarkal** [comment]
Two things come to mind:

1. In the HTTP headers from the first request you should have `link` headers telling you how many pages actually exist to prevent you from querying a non existent page.
```
Link →< http://54.67.43.161:8080/api/users?page=2&perPage=10 >; rel="next",< http://54.67.43.161:8080/api/users?page=1&perPage=10 >; rel="prev",< http://54.67.43.161:8080/api/users?page=1&perPage=10 >; rel="first",< http://54.67.43.161:8080/api/users?page=2&perPage=10 >; rel="last",< http://54.67.43.161:8080/api/users?page=18&perPage=1 >; rel="final"
```

2. Can you compare the response time to just using `perPage=500`?

**Comment by ActionCactus** [comment]
Just using perPage=500, with no page specified, goes quickly.  Specifying a page=2 makes things go slowly.

**Comment by tarkal** [comment]
Do you mean `page=2&perPage=500` goes slowly?

**Comment by ActionCactus** [comment]
Yes, much slower than page=1&perPage=500

**Comment by tarkal** [comment]
I can confirm that the query time is dramatically increased with a proportional relation to the number of items requested in the empty collection. 

I will investigate this and make it part of the query optimisation (#70) issue. In the short term I would recommend that you either use the simple solution and just specify `perPage=500` without using the `page` query. There are currently no nodes that have anywhere near 500 properties so the request would be safe in terms of retrieving all the properties.

The slightly more involved solution (which we use in the GUI). Is to use the response headers to perform the queries.
